Improve CI/CD

Chris2fourlaw opened the pull request:
Notes:
- Runs on all pushes and pull requests
- Separate Integration job from Deployment job
  - Run integration always, skip deployment if not master or beta branch
- Integration runs on latest and latest LTS of NodeJS (currently 18.x and 16.x respectively)
- Deployment runs on latest LTS (16.x currently)

mfish33 commented:
This looks like a nice improvement! Just two comments:
1. Is there an improvement to have the workflow run on push and pull_request since it seems like push is just a strict superset to me?
2. The build seems to be failing with node 18.X. If you would like to take the time to investigate and fix this, feel free. Otherwise, I am ok with just using node LTS.

Chris2fourlaw commented:
1. Without the pull_request workflow, we won't know if someone's PR changes are breaking until it is merged
2. I think I fixed it, I rebased my commit onto the beta branch and reran

mfish33 commented:
Oh, I see the reason for the pull_requests option! We just have been doing everything inside the repo so far. The reason why the integration and deploy job was all in one is that when the jobs are separated you must completely rebuild the monorepo. Since we are adding both lts and latest as a matrix I am ok with this change but the deployment job must also run npm install and build.

Chris2fourlaw commented:
Oh good catch. I’ll dig and check if the artifacts are saved between jobs

Chris2fourlaw commented:
After digging, I found https://github.com/actions/upload-artifact but I realized this whole job split is kinda useless abstraction.

The updated changes this PR does is:
- Trigger on PRs and test against post-merge
- Cache between CI/CD runs (see https://github.com/actions/setup-node#caching-global-packages-data)
- CI against LTS and latest
- CD against LTS
